lib/thread.c

```diff
@@ -1715,17 +1722,35 @@ void thread_call(struct thread *thread)
 				 memory_order_seq_cst);
 
 #ifdef CONSUMED_TIME_CHECK
-	if (realtime > CONSUMED_TIME_CHECK) {
+	if (cputime > CONSUMED_TIME_CHECK) {
 		/*
-		 * We have a CPU Hog on our hands.
+		 * We have a CPU Hog on our hands.  The time FRR
+		 * has spent doing actual work ( not sleeping )
+		 * is greater than 5 seconds.
 		 * Whinge about it now, so we're aware this is yet another task
 		 * to fix.
 		 */
+		atomic_fetch_add_explicit(&thread->hist->total_cpu_warn,
+					  1, memory_order_seq_cst);
 		flog_warn(
-			EC_LIB_SLOW_THREAD,
+			EC_LIB_SLOW_THREAD_CPU,
 			"SLOW THREAD: task %s (%lx) ran for %lums (cpu time %lums)",
 			thread->xref->funcname, (unsigned long)thread->func,
 			realtime / 1000, cputime / 1000);
+	} else if (realtime > CONSUMED_TIME_CHECK) {
+		/*
```

Comment:
So now we are using the condition that was earlier used to detect a CPU Hog to detect a CPU starvation? That doesn't seem to look good.

Reply:
I don't see how this will detect a CPU starvation. Can you please explain with an example?

Reply:
I never claim it's cpu starvation.  I'm claiming it's CPU bound or Wall bound.  It is up to the operator to look at it and make a determination.

But given that let's consider some things:

realtime is always going to be equal to or larger than cpu time.
I wanted to quickly to be able to determine which one is happening without a bunch of code.
Dividing the current situation into two choices allows me to make a determination of what is happening in a generic sense on the system, with more granularity than we have now.

As a developer I am more interested in the CPU bound process taking over 5 seconds than I am Wall time processes taking over CPU time( but the reality is it would be nice to look at the log message and understand it better )

If I am interested in starvation.  I would use a simple formula along the lines of if cpu time < .5 second and realtime-cputime is greater than X ( say 2 seconds ) we probably have a starvation situation.

Reply:
I'll just chime in: I do think this is useful, potentially, and in combination with Donald's other PR that is more timer-specific, I think it will help identify environment/OS and internal issue.